### PR TITLE
access_log: add %PROXY_PROTOCOL_TLVS(TYPE)% substitution string

### PR DESCRIPTION
**Pull request by betsybutton**
Commit Message: add %PROXY_PROTOCOL_TLVS(TYPE)% substitution string to extract proxy protocol TLVs stored in the filter state with the given type
Additional Description: feature change was discussed in https://github.com/envoyproxy/envoy/issues/19046
Risk Level: low
Testing: unit tests, manual testing
Docs Changes: new substitution string added to `docs/root/configuration/observability/access_log/usage.rst1`
Release Notes: added to `changelogs/current.yaml`
Platform Specific Features: none
Fixes https://github.com/envoyproxy/envoy/issues/19046

### CODE REVIEW

File: source/common/formatter/stream_info_formatter.cc [hunk 1]
```diff
@@ -248,6 +250,56 @@ FilterStateFormatter::formatValue(const StreamInfo::StreamInfo& stream_info) con
...
+    throw EnvoyException(fmt::format("Invalid parameter provided for PROXY_PROTOCOL_TLVS header: "
+                                     "{}. Must be a positive integer less than 256.",
+                                     tlv_type_str));
+  }
+}
+
+absl::optional<std::string> ProxyProtocolTlvsFormatter::format(const StreamInfo::StreamInfo& stream_info) const {
+  const auto& typed_state =
+      stream_info.filterState().getDataReadOnly<Network::ProxyProtocolFilterState>(
+          Network::ProxyProtocolFilterState::key());
```
Comment (wbpcode) [inline source/common/formatter/stream_info_formatter.cc:273]: nit: `const auto*`. The typed_state is a pointer rather than a ref.

**Review by wbpcode — COMMENTED** [review]
Thanks for your contribution. But could you check the https://github.com/envoyproxy/envoy/pull/29223 first. I think it has provided a common way to extract sub attributes from filter state object.
Although by that way, we cannot check the correctness of the key at configuration phase. But it just a minor drawback for access log.

/wait

**Comment by betsybutton** [comment]
Hi @wbpcode, thanks for taking a look. https://github.com/envoyproxy/envoy/pull/29223 could solve the question of serializing the TLVs, but I'd prefer a unique substitution formatter to handle this use case because it allows us to filter out unwanted TLV types, as defined in the config. Could we continue to pursue merging these changes?

**Comment by wbpcode** [comment]
> but I'd prefer a unique substitution formatter to handle this use case because it allows us to filter out unwanted TLV types, as defined in the config.

I think #29223 could filter unwanted TLV types out also? 

> Could we continue to pursue merging these changes?

Sorry, but I was not convinced by your current reason.  IMO, the #29223 should be a final solution to log filter state objects or their sub attributes. I still didn't see the necessity to add the new command to the core code.

**Comment by betsybutton** [comment]
True, it is possible for us to accomplish this with the general formatter. I'll start work on the required changes for proxy protocol to use the field-level formatter and open a separate PR when they're ready. Thanks @wbpcode!

**Pull request closed without merging**